include/data/representation_converter.hpp

```diff
@@ -0,0 +1,257 @@
+   *
+   * @tparam SourceType The source representation type (must derive from idata_representation)
+   * @tparam TargetType The target representation type (must derive from idata_representation)
+   * @param converter The converter function to register
+   * @throws std::runtime_error if a converter for this type pair already exists
+   */
+  template <typename SourceType, typename TargetType>
+  void register_converter(representation_converter_fn converter)
+  {
+    static_assert(std::is_base_of_v<idata_representation, SourceType>,
+                  "SourceType must derive from idata_representation");
+    static_assert(std::is_base_of_v<idata_representation, TargetType>,
+                  "TargetType must derive from idata_representation");
+
+    converter_key key{std::type_index(typeid(SourceType)), std::type_index(typeid(TargetType))};
+    register_converter_impl(key, std::move(converter));
+  }
+
+  /**
+   * @brief Check if a converter exists for the given type pair.
+   *
+   * @tparam SourceType The source representation type
+   * @tparam TargetType The target representation type
+   * @return true if a converter is registered, false otherwise
+   */
+  template <typename SourceType, typename TargetType>
+  bool has_converter() const
+  {
+    converter_key key{std::type_index(typeid(SourceType)), std::type_index(typeid(TargetType))};
+    return has_converter_impl(key);
+  }
+
+  /**
+   * @brief Check if a converter exists for the given source instance and target type.
+   *
+   * @tparam TargetType The target representation type
+   * @param source The source data representation
+   * @return true if a converter is registered, false otherwise
+   */
+  template <typename TargetType>
+  bool has_converter_for(const idata_representation& source) const
+  {
+    converter_key key{std::type_index(typeid(source)), std::type_index(typeid(TargetType))};
+    return has_converter_impl(key);
+  }
+
+  /**
+   * @brief Convert a data representation to another type.
+   *
+   * @tparam TargetType The target representation type
+   * @param source The source data representation
+   * @param target_memory_space The target memory space for the new representation
+   * @param stream CUDA stream for memory operations
+   * @return std::unique_ptr<TargetType> The converted representation
+   * @throws std::runtime_error if no converter is registered for the type pair
+   *
+   * @example
+   * auto result = registry.convert<gpu_table_representation>(source, space);
+   */
+  template <typename TargetType>
+  std::unique_ptr<TargetType> convert(idata_representation& source,
+                                      const memory::memory_space* target_memory_space,
+                                      rmm::cuda_stream_view stream = rmm::cuda_stream_default) const
+  {
+    converter_key key{std::type_index(typeid(source)), std::type_index(typeid(TargetType))};
+    auto result = convert_impl(key, source, target_memory_space, stream);
```

Comment:
Maybe there should be a check in case the converter is found but the the target memory space is invalid?

For example if someone converting gpu_data_representation to host_data_representation but targeting another GPU memory space, it should be invalid